azure-iot-device/azure/iot/device/common/pipeline/pipeline_thread.py

```diff
@@ -0,0 +1,199 @@
+  as the the "callback thread".  This is not meant for callbacks into pipeline
+  code.  Those callbacks should still execute on the pipeline thread.  The
+  `invoke_on_callback_thread_nowait` decorator is used to ensure that callbacks
+  execute on the callback thread.
+
+4. Decorators which cause thread switches are used only when necessary.  The
+  pipeline thread is only entered in places where we know that external code is
+  calling into the pipeline (such as a client API call or a callback from a
+  third-party library).  Likewise, the callback thread is only entered in places
+  where we know that the pipeline is calling back into client code.
+
+5. Exceptions raised from the pipeline thread are still able to be caught by
+  the function which entered the pipeline thread.
+
+5. Calls into the pipeline thread can either block or not block.  Blocking is used
+  for cases where the caller needs a return value from the pipeline or is
+  expecting to handle any errors raised from the pipeline thread.  Blocking is
+  not used when the code calling into the pipeline is not waiting for a response
+  and is not expecting to handle any exceptions, such as protocol library
+  handlers which call into the pipeline to deliver protocol messages.
+
+6. Calls into the callback thread could theretically block, but we currently
+  only have decorators which enter the callback thread without blocking.  This
+  is done to ensure that client code does not execute on the pipeline thread and
+  also to ensure that the pipline thread is not blocked while waiting for client
+  code to execute.
+
+These decorators use concurrent.futures.Future and the ThreadPoolExecutor because:
+
+1. The thread pooling with a pool size of 1 gives us a single thread to run all
+  pipeline operations and a different (single) thread to run all callbacks.  If
+  the code attempts to run a second pipeline operation (or callback) while a
+  different one is running, the ThreadPoolExecutor will queue the code until the
+  first call is completed.
+
+2. The concurent.futures.Future object properly handles both Exception and
+  BaseException errors, re-raising them when the Future.result method is called.
+  threading.Thread.get() was not an option because it doesn't re-raise
+  BaseException errors when Thread.get is called.
+
+3. concurrent.futures is available as a backport to 2.7.
+
+"""
+
+
+def _get_named_executor(thread_name):
+    """
+    Get a ThreadPoolExecutor object with the given name.  If no such executor exists,
+    this function will create on with a single worker and assign it to the provided
+    name.
+    """
+    executor = getattr(_get_named_executor, thread_name, None)
+    if not executor:
+        logger.info("Creating {} executor".format(thread_name))
+        executor = ThreadPoolExecutor(max_workers=1)
+        setattr(_get_named_executor, thread_name, ThreadPoolExecutor(max_workers=1))
+    return executor
+
+
+def _get_thread_local_storage():
```

Comment:
>_get_thread_local_storage [](start = 4, length = 25)

if this just returns thread specific local data is there a reason why it is named `local_storage` ?

Reply:
"thread local storage" is an industry term.  Regardless, I've removed this in my next iteration.

---
In reply to: [301802099](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r301802099) [](ancestors = 301802099)